src/mlpack/core/metrics/iou_metric_impl.hpp

```diff
@@ -0,0 +1,57 @@
+/**
+ * @file iou_metric_impl.hpp
+ * @author Kartik Dutt
+ *
+ * Implementation of Intersection Over Union metric.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_METRICS_IOU_IMPL_HPP
+#define MLPACK_CORE_METRICS_IOU_IMPL_HPP
+
+// In case it hasn't been included.
+#include "iou_metric.hpp"
+
+namespace mlpack {
+namespace metric {
+
+template<bool useCoordinates>
+template <typename VecTypeA, typename VecTypeB>
+typename VecTypeA::elem_type IoU<useCoordinates>::Evaluate(
+    const VecTypeA& a,
+    const VecTypeB& b)
+{
+  Log::Assert(a.n_elem == b.n_elem && a.n_elem == 4, "Incorrect \
+      shape for bounding boxes. They must contain 4 elements either be \
+      (x0, y0, x1, y1) or (x0, y0, h, w). Refer to the documentation \
+      for more information.");
```

Comment:
Nice!, would it be a good idea to check if the coordinates consist of a rectangle?

Reply:
Yes, that makes sense when coordinates are represented as `(x0, y0, x1, y1)` incase they form a line or their order isn't correct.